pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/MailboxContentStreamObserver.java

```diff
@@ -93,7 +105,19 @@ public void onNext(Mailbox.MailboxContent mailboxContent) {
 
     if (!mailboxContent.getMetadataMap().containsKey(ChannelUtils.MAILBOX_METADATA_BEGIN_OF_STREAM_KEY)) {
       // when the receiving end receives a message put it in the mailbox queue.
-      _receivingBuffer.offer(mailboxContent);
+      // TODO: pass a timeout to _receivingBuffer.
+      if (!_receivingBuffer.offer(mailboxContent)) {
```

Comment:
question: 
1. so when `_receivingBuffer.offer(mailboxContent) == false` this means the buffer somehow cannot be done correctly, yes?
2.  do we know what's the timeout behavior is? (is adding a timeout on the `offer`method going to block the onNext() thread and thus block the next GRPC call and eventually creates back pressure? (CC @agavra if you have any thoughts on this)

Reply:
from the documentation:
```
This API is asynchronous, so methods may return before the operation
 completes. The API provides no guarantees for how quickly an operation 
will complete, so utilizing flow control via ClientCallStreamObserver 
and ServerCallStreamObserver to avoid excessive buffering is recommended 
for streaming RPCs. gRPC's implementation of onError() on client-side 
causes the RPC to be cancelled and discards all messages, so completes 
quickly.

gRPC guarantees it does not block on I/O in its implementation, 
but applications are allowed to perform blocking operations in 
their implementations. However, doing so will delay other callbacks 
because the methods cannot be called concurrently.
```

If we block here, I think GRPC will just buffer on its end

Reply:
yes. Eventually, we should utilize GRPC streaming buffer for this.  We need to put more thoughts into that. As said in the PR description , this PR doesn't intent to fix the issue. It serves as  a quick bandaid to throw error instead of silently returning incorrect result.

Reply:
yes +1, I agree with this PR

Reply:
+1